File: packages/react-native-vision-camera-text-recognition/android/src/main/java/com/margelo/nitro/camera/textrecognition/HybridTextRecognizer.kt
package com.margelo.nitro.camera.textrecognition

import androidx.annotation.OptIn
import androidx.camera.core.ExperimentalGetImage
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.textrecognition.extensions.toInputImage
import com.margelo.nitro.core.Promise

class HybridTextRecognizer : HybridTextRecognizerSpec() {
  private val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)

  @OptIn(ExperimentalGetImage::class)
  override fun recognizeText(frame: HybridFrameSpec): RecognizedText {
    val image = frame.toInputImage()
    val task = recognizer.process(image)
    val text = Tasks.await(task)
    return text.toRecognizedText()
  }

  override fun recognizeTextAsync(frame: HybridFrameSpec): Promise<RecognizedText> {
    val promise = Promise<RecognizedText>()

    val image = frame.toInputImage()
    recognizer
      .process(image)
      .addOnSuccessListener { text ->
        promise.resolve(text.toRecognizedText())
      }.addOnFailureListener { error ->
        promise.reject(error)
      }

    return promise
  }
}

File: packages/react-native-vision-camera-text-recognition/ios/HybridTextRecognizer.swift
//
//  HybridTextRecognizer.swift
//  VisionCameraTextRecognition
//

import MLKitTextRecognition
import MLKitVision
import NitroModules
import VisionCamera

class HybridTextRecognizer: HybridTextRecognizerSpec {
  private let recognizer: TextRecognizer

  override init() {
    self.recognizer = TextRecognizer.textRecognizer(options: TextRecognizerOptions())
    super.init()
  }

  func recognizeText(frame: any HybridFrameSpec) throws -> RecognizedText {
    let image = try frame.toMLImage()
    let text = try recognizer.results(in: image)
    return text.toRecognizedText()
  }

  func recognizeTextAsync(frame: any HybridFrameSpec) throws -> Promise<RecognizedText> {
    let promise = Promise<RecognizedText>()

    let image = try frame.toMLImage()
    recognizer.process(image) { text, error in
      if let error {
        promise.reject(withError: error)
        return
      }
      if let text {
        promise.resolve(withResult: text.toRecognizedText())
      } else {
        promise.resolve(withResult: RecognizedText(text: "", blocks: []))
      }
    }

    return promise
  }
}
